java-checks/src/main/java/org/sonar/java/checks/EqualsNotOverriddenInSubclassCheck.java

```diff
@@ -0,0 +1,123 @@
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import org.sonar.api.server.rule.RulesDefinition;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.java.model.AbstractTypedTree;
+import org.sonar.java.model.declaration.ClassTreeImpl;
+import org.sonar.java.resolve.Symbol;
+import org.sonar.java.resolve.Symbol.MethodSymbol;
+import org.sonar.java.resolve.Symbol.TypeSymbol;
+import org.sonar.java.resolve.Type;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.Modifier;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.Tree.Kind;
+import org.sonar.plugins.java.api.tree.VariableTree;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.List;
+
+@Rule(
+  key = "S2160",
+  name = "Subclasses that add fields should override \"equals\"",
+  tags = {"bug"},
+  priority = Priority.MAJOR)
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.LOGIC_RELIABILITY)
+@SqaleConstantRemediation("30min")
+public class EqualsNotOverriddenInSubclassCheck extends SubscriptionBaseVisitor {
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    return ImmutableList.of(Kind.CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    ClassTree classTree = (ClassTree) tree;
+    if (hasSemantic()
+      && hasAtLeastOneField(classTree)
+      && !implementsEquals(classTree)
+      && parentClassImplementsEquals(classTree)) {
+      addIssue(classTree, "Override this superclass' \"equals\" method.");
+    }
+  }
+
+  private boolean hasAtLeastOneField(ClassTree classTree) {
+    for (Tree member : classTree.members()) {
+      if (isField(member)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private boolean isField(Tree tree) {
+    return tree.is(Kind.VARIABLE) && !((VariableTree) tree).modifiers().modifiers().containsAll(Lists.newArrayList(Modifier.FINAL, Modifier.STATIC));
+  }
+
+  private boolean implementsEquals(ClassTree classTree) {
+    return hasEqualsMethod(((ClassTreeImpl) classTree).getSymbol());
+  }
+
+  private boolean parentClassImplementsEquals(ClassTree tree) {
+    Tree superClass = tree.superClass();
+    if (superClass != null) {
+      TypeSymbol superClassSymbol = ((AbstractTypedTree) superClass).getSymbolType().getSymbol();
+      while (!superClassSymbol.getType().is("java.lang.Object")) {
+        if (hasEqualsMethod(superClassSymbol)) {
+          return true;
+        }
+        Type superClassType = superClassSymbol.getSuperclass();
+        if (superClassType == null) {
```

Comment:
Why this null check ? it deserves at least a comment as this case is not supposed to happen.